Documentation/ceph-csi-drivers.md

```diff
@@ -0,0 +1,382 @@
+default     pod/csi-rbdplugin-attacher-0               1/1     Running     0          132m
+default     pod/csi-rbdplugin-provisioner-0            1/1     Running     0          132m
+
+NAMESPACE   NAME                                TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)         AGE
+default     service/csi-rbdplugin-attacher      ClusterIP   10.107.65.227    <none>        12345/TCP       132m
+default     service/csi-rbdplugin-provisioner   ClusterIP   10.108.120.159   <none>        12345/TCP       132m
+
+NAMESPACE   NAME                             DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
+default     daemonset.apps/csi-rbdplugin     1         1         1       1            1           <none>          132m
+
+NAMESPACE   NAME                                         DESIRED   CURRENT   AGE
+default     statefulset.apps/csi-rbdplugin-attacher      1         1         132m
+default     statefulset.apps/csi-rbdplugin-provisioner   1         1         132m
+
+```
+You should see an output similar to this when you run `kubectl get all`.
+
+## Test the CSI Driver
+
+Once the plugin is successfully deployed, test it by running the following example
+
+This example is based on the ceph-csi [examples](https://github.com/ceph/ceph-csi/tree/master/examples) directory. It will describe how to test the RBD csi-driver, but it is very similar to run the cephfs one.
+
+### Create the StorageClass:
+##### storageclass.yaml
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+   name: csi-rbd
+provisioner: csi-rbdplugin 
+parameters:
+    # Comma separated list of Ceph monitors
+    # if using FQDN, make sure csi plugin's dns policy is appropriate.
+    monitors: mon1:port,mon2:port,...
+
+    # if "monitors" parameter is not set, driver to get monitors from same
+    # secret as admin/user credentials. "monValueFromSecret" provides the
+    # key in the secret whose value is the mons
+    #monValueFromSecret: "monitors"
+
+    
+    # Ceph pool into which the RBD image shall be created
+    pool: rbd
+
+    # RBD image format. Defaults to "2".
+    imageFormat: "2"
+
+    # RBD image features. Available for imageFormat: "2". CSI RBD currently supports only `layering` feature.
+    imageFeatures: layering
+    
+    # The secrets have to contain Ceph admin credentials.
+    csiProvisionerSecretName: csi-rbd-secret
+    csiProvisionerSecretNamespace: default
+    csiNodePublishSecretName: csi-rbd-secret
+    csiNodePublishSecretNamespace: default
+
+    # Ceph users for operating RBD
+    adminid: admin
+    userid: kubernetes
+    # uncomment the following to use rbd-nbd as mounter on supported nodes
+    #mounter: rbd-nbd
+reclaimPolicy: Delete
+```
+### Create the Secret:
+##### secret.yaml
+```
+apiVersion: v1
+kind: Secret
+metadata:
+  name: csi-rbd-secret
+  namespace: default 
+data:
+  # Key value corresponds to a user name defined in ceph cluster
+  admin: BASE64-ENCODED-PASSWORD
```

Comment:
you can find the secret in the `rook-ceph-mon` secret of the `rook-ceph` namespace. Can we re-use that instead of creating a new secret? Or if we need the secret in the default namespace we could at least use that secret to copy to the other secret.

Reply:
It would be great to reuse it but I haven't found a solution for this; the `rook-ceph-mon` contains an `admin-secret` key as you can see here:
```
$ kubectl describe secret rook-ceph-mon -n rook-ceph
Name:         rook-ceph-mon
Namespace:    rook-ceph
Labels:       <none>
Annotations:  <none>

Type:  kubernetes.io/rook

Data
====
admin-secret:  40 bytes
cluster-name:  9 bytes
fsid:          36 bytes
mon-secret:    40 bytes
```
but the ceph-csi plugin expect the key to be the username of the ceph user, like this:
```
$ kubectl describe secret csi-rbd-secret
Name:         csi-rbd-secret
Namespace:    default
Labels:       <none>
Annotations:  <none>

Type:  Opaque

Data
====
admin:  40 bytes
```
Otherwise the plugin throws errors like these: 
``` 
W1024 12:47:59.861327       1 controllerserver.go:130] failed to create volume: RBD key for ID: admin not found
```
and won't create the PVC.

I don't think the secret _has_ to be in the default namespace since there's `csiProvisionerSecretNamespace` parameter to specify it so it shouldn't a problem. If you have an idea how to handle this I'm all ears!

Reply:
Ok, I see that what I was suggesting wouldn't help. If there were already a secret existing in the rook-ceph namespace I was suggesting that it could be extracted with a command such as the following:
```
kubectl -n rook-ceph get secret rook-ceph-mon -o yaml | grep admin-secret | awk '{print $2}' | base64 --decode
```
But since that's a different secret, sounds like they would need to use `ceph auth ls` as suggested below.